pkg/morph/event/notary_preparator.go

```diff
@@ -0,0 +1,243 @@
+)
+
+var (
+	errNotContractCall            = errors.New("received main tx is not a contract call")
+	errUnexpectedWitnessAmount    = errors.New("received main tx has unexpected amount of signers")
+	errIncorrectProxyWitnesses    = errors.New("received main tx has non-empty Proxy witnesses")
+	errIncorrectAlphabet          = errors.New("received main tx has incorrect Alphabet verification")
+	errIncorrectNotaryPlaceholder = errors.New("received main tx has incorrect Notary contract placeholder")
+	errIncorrectAttributesAmount  = errors.New("received main tx has incorrect attributes amount")
+	errIncorrectAttribute         = errors.New("received main tx has incorrect attribute")
+	errUnexpectedOpcode           = errors.New("received main tx has unexpected(not PUSHDATA) NeoVM opcodes")
+
+	// ErrTXAlreadyHandled is returned if received TX has already been signed.
+	ErrTXAlreadyHandled = errors.New("received main tx has already been handled")
+)
+
+// PreparatorPrm groups the required parameters of the Preparator constructor.
+type PreparatorPrm struct {
+	alphaKeys client.AlphabetKeys
+}
+
+// Preparator implements NotaryPreparator interface.
+type Preparator struct {
+	// contractSysCall contract call in NeoVM
+	contractSysCall []byte
+	// dummyInvocationScript is invocation script from TX that is not signed.
+	dummyInvocationScript []byte
+
+	alphaKeys client.AlphabetKeys
+}
+
+// notaryPreparator inits and returns NotaryPreparator.
+//
+// Considered to be used for preparing notary request
+// for parsing it by event.Listener.
+func notaryPreparator(prm PreparatorPrm) NotaryPreparator {
+	switch {
+	case prm.alphaKeys == nil:
+		panic("alphabet keys source must not be nil")
+	}
+
+	contractSysCall := make([]byte, 4)
+	binary.LittleEndian.PutUint32(contractSysCall, interopnames.ToID([]byte(interopnames.SystemContractCall)))
+
+	dummyInvocationScript := append([]byte{byte(opcode.PUSHDATA1), 64}, make([]byte, 64)...)
+
+	return Preparator{
+		contractSysCall:       contractSysCall,
+		dummyInvocationScript: dummyInvocationScript,
+		alphaKeys:             prm.alphaKeys,
+	}
+}
+
+// Prepare converts raw notary requests to NotaryEvent.
+//
+// Returns ErrTXAlreadyHandled if transaction shouldn't be
+// parsed and handled. It is not "error case". Every handled
+// transaction is expected to be received one more time
+// from the Notary service but already signed. This happens
+// since every notary call is a new notary request in fact.
+func (p Preparator) Prepare(nr *payload.P2PNotaryRequest) (NotaryEvent, error) {
+	currentAlphabet, err := p.alphaKeys()
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch Alphabet public keys: %w", err)
+	}
+
+	// validate main TX's notary attribute
+	err = p.validateAttributes(nr.MainTransaction.Attributes, currentAlphabet)
+	if err != nil {
+		return nil, err
+	}
+
+	// validate main TX's witnesses
+	err = p.validateWitnesses(nr.MainTransaction.Scripts, currentAlphabet)
```

Comment:
1. such checks as: exactly 3 signers; second is multiacc of Alphabet? the first one is always proxy(does not need to check, I guess, since it is filtered in subscription process), and the third one is always notary(does not need to check, I guess, since this check is done with notary service), am I correct?
2. agree, will add such check

Reply:
Yes, that's true.